src/dwn/models/record.js

```diff
@@ -0,0 +1,244 @@
+    if (options?.data) {
+      // this.#readableStream = isReadableWebStream(options.data) ? toIsomorphicReadableStream(options.data) : options.data;
+      this.#readableStream = isReadableWebStream(options.data) ? new ReadableWebToNodeStream(options.data) : options.data;
+    }
+  }
+
+  // Mutable Web5 Record Class properties.
+  get author() { return this.#author; }
+  get isFrozen() { return this.#isFrozen; }
+  get target() { return this.#target; }
+  set author(author) { this.#author = author; }
+  set target(target) { this.#target = target; }
+  
+  // Immutable DWN Record properties.
+  get id() { return this.#recordId; }
+  get contextId() { return this.#contextId; }
+  get dataFormat() { return this.#descriptor?.dataFormat; }
+  get dateCreated() { return this.#descriptor?.dateCreated; }
+  get interface() { return this.#descriptor?.interface; }
+  get method() { return this.#descriptor?.method; }
+  get parentId() { return this.#descriptor?.parentId; }
+  get protocol() { return this.#descriptor?.protocol; }
+  get recipient() { return this.#descriptor?.recipient; }
+  get schema() { return this.#descriptor?.schema; }
+
+  // Mutable DWN Record properties.
+  get dataCid() { return this.#descriptor?.dataCid; }
+  get dataSize() { return this.#descriptor?.dataSize; }
+  get dateModified() { return this.#descriptor?.dateModified; }
+  get datePublished() { return this.#descriptor?.datePublished; }
+  get published() { return this.#descriptor?.published; }
+
+  /**
+   * Data handling.
+   */
+
+  get data() {
+    if (!this.#encodedData && !this.#readableStream) {
+      // `encodedData` will be set if `dataSize` <= DwnConstant.maxDataSizeAllowedToBeEncoded. (10KB as of April 2023)
+      // `readableStream` will be set if Record was instantiated from a RecordsRead reply.
+      // If neither of the above are true, then the record must be fetched from the DWN.
+      this.#readableStream = this.#dwn.records.read(this.#target, { author: this.#author, message: { recordId: this.#recordId } })
+        .then((response) => response.record )
+        .then((record) => { return record.data; });
+    }
+
+    if (typeof this.#encodedData === 'string') {
+      // If `encodedData` is set, then it is expected that:
+      // `dataSize` <= DwnConstant.maxDataSizeAllowedToBeEncoded (10KB as of April 2023)
+      // type is Uint8Array bytes if the Record object was instantiated from a RecordsWrite response
+      // type is Base64 URL encoded string if the Record object was instantiated from a RecordsQuery response
+      // If it is a string, we need to Base64 URL decode to bytes
+      this.#encodedData = Encoder.base64UrlToBytes(this.#encodedData);
+    }
+
+    const self = this; // Capture the context of the `Record` instance.
+    const dataObj = {
+      async json() {
+        if (self.#encodedData) return this.text().then(JSON.parse);
+        if (self.#readableStream) return this.text().then(JSON.parse);
```

Comment:
it's really odd that we use `self` and then `this`.  can we rework this to make it less confusing?  ditto below

also, can we combine these into a single `if`?

Reply:
The `self` is necessary as currently written so that the functions within the `dataObj` can access both the instance `this` and the other methods within the object (e.g., `this.text().then(JSON.parse`).

If we refactor the approach then it may be possible but this wouldn't change the functionality.

Opened #46 to track considering improvements to the approach